crates/bevy_input/src/gamepad.rs

```diff
@@ -1127,102 +983,256 @@ impl ButtonAxisSettings {
+                let gamepad_button = GamepadButton::new(gamepad, *button_type);
+                button_input.reset(gamepad_button);
+                button_axis.set(gamepad_button, 0.0);
+            }
+            for axis_type in &ALL_AXIS_TYPES {
+                axis.set(GamepadAxis::new(gamepad, *axis_type), 0.0);
             }
+        } else {
+            gamepads.deregister(gamepad);
+            info!("{:?} Disconnected", gamepad);
 
-            GamepadEventType::Disconnected => {
-                gamepads.deregister(event.gamepad);
-                info!("{:?} Disconnected", event.gamepad);
+            for button_type in &ALL_BUTTON_TYPES {
+                let gamepad_button = GamepadButton::new(gamepad, *button_type);
+                button_input.reset(gamepad_button);
+                button_axis.remove(gamepad_button);
+            }
+            for axis_type in &ALL_AXIS_TYPES {
+                axis.remove(GamepadAxis::new(gamepad, *axis_type));
             }
-            _ => (),
         }
     }
 }
 
-/// Modifies the gamepad resources and sends out gamepad events.
-///
-/// The resources [`Input<GamepadButton>`], [`Axis<GamepadAxis>`], and [`Axis<GamepadButton>`] are updated
-/// and the [`GamepadEvent`]s are sent according to the received [`GamepadEventRaw`]s respecting the [`GamepadSettings`].
-///
-/// ## Differences
+#[derive(Debug, Clone, PartialEq, Reflect, FromReflect)]
+#[reflect(Debug, PartialEq)]
+#[cfg_attr(
+    feature = "serialize",
+    derive(serde::Serialize, serde::Deserialize),
+    reflect(Serialize, Deserialize)
+)]
+pub enum GamepadConnection {
+    Connected(GamepadInfo),
+    Disconnected,
+}
+
+/// A Gamepad connection event. Created when a connection to a gamepad
+/// is established and when a gamepad is disconnected.
+#[derive(Debug, Clone, PartialEq, Reflect, FromReflect)]
+#[reflect(Debug, PartialEq)]
+#[cfg_attr(
+    feature = "serialize",
+    derive(serde::Serialize, serde::Deserialize),
+    reflect(Serialize, Deserialize)
+)]
+pub struct GamepadConnectionEvent {
+    /// The gamepad whose connection status changed.
+    pub gamepad: Gamepad,
+    /// The change in the gamepads connection.
+    pub connection: GamepadConnection,
+}
+
+impl GamepadConnectionEvent {
+    pub fn new(gamepad: Gamepad, connection: GamepadConnection) -> Self {
+        Self {
+            gamepad,
+            connection,
+        }
+    }
+
+    pub fn connected(&self) -> bool {
+        matches!(self.connection, GamepadConnection::Connected(_))
+    }
+
+    pub fn disconnected(&self) -> bool {
+        !self.connected()
+    }
+}
+
+/// Gamepad axis changed event, used internally to create a [`GamepadAxisChangedEvent`]
+/// and update `Input<T>` resources in the [`gamepad_raw_axis_event_system`]. The
+/// difference between [`RawGamepadAxisChangedEvent`] and [`GamepadAxisChangedEvent`]
+/// is that the latters axis "value" is filtered, based on a [`GamepadSettings`].
```

Comment:
Nit: The first line should really just be a quick description of the item. Further details, should be in a separate paragraph.
```suggestion
/// Gamepad axis changed event, used internally to create a [`GamepadAxisChangedEvent`]
/// and update `Input<T>` resources in the [`gamepad_raw_axis_event_system`]. 
///
/// The difference between [`RawGamepadAxisChangedEvent`] and [`GamepadAxisChangedEvent`]
/// is that the latters axis "value" is filtered, based on a [`GamepadSettings`].
```